src/motion_model.py

```diff
@@ -1,225 +1,246 @@
+    def reflect(self, old_pt, new_pt, index) -> tuple:
+        norm_v = norm(self.velocities[index])
+        theta = self.boundary.reflect_velocity(old_pt, new_pt)
+        self.velocities[index] = (norm_v*cos(theta), norm_v*sin(theta))
+        return self.boundary.reflect_point(old_pt, new_pt)
+
+    ## The potential function is a sum of exponential terms that takes the negative pairwise distances as an argument.
+    # Each summand is a difference of two exponential terms, with the "attraction" term being subtracted from the
+    # "repulsion" term for each given pairwise distance. (This is so that the gradient of this function has the
+    # repulsion term being subtracted from the attraction term)
+    # The coefficients of the repulsion and attraction terms determine the strength of repulsion and attraction,
+    # respectively, at a given distance.
+    # /f$ /sum{C_re^{-|x_i - x_j|/L_r} - C_ae^{-|x_i - x_j|/L_a}} /f$
+    def gradient(self, xs, ys):
+        gradUx, gradUy = [0.0] * self.n_sensors, [0.0] * self.n_sensors
+
+        for i in range(0, self.n_sensors):
+            for j in range(self.n_sensors):
+                r = norm((xs[i] - xs[j], ys[i] - ys[j]))
+                if 0.0 < r < 2 * self.sensing_radius:
+
+                    attract_term = (self.DO_coeff[0]*np.exp(-r/self.DO_coeff[1]) / (self.DO_coeff[1]*r))
+                    repel_term = (self.DO_coeff[2]*np.exp(-r/self.DO_coeff[3]) / (self.DO_coeff[3]*r))
+
+                    gradUx[i] += (xs[i]-xs[j]) * attract_term - (xs[i]-xs[j]) * repel_term
+                    gradUy[i] += (ys[i]-ys[j]) * attract_term - (ys[i]-ys[j]) * repel_term
+
+        return array(gradUx), array(gradUy)
+
+    ## This computes dx/dt, dy/dt, dvx/dt, dvy/dt which are later solved to compute the next set of x,y coordinates
+    def time_derivative(self, _, state):
+        # ode solver gives us np array in the form [xvals | yvals | vxvals | vyvals]
+        # split into individual np array
+        split_state = [state[i:i + self.n_sensors] for i in range(0, len(state), self.n_sensors)]
+        gradU = self.gradient(split_state[0], split_state[1])
+
+        # Need to compute time derivative of each,
+        # I just have d(x, y)/dt = (vx, vy), d(vx, vy)/dt = (1, -1)
+        dxdt = split_state[2]
+        dydt = split_state[3]
+        dvxdt = (5 - (0.5 * norm((dxdt, dydt))**2)) * dxdt - gradU[0]
+        dvydt = (5 - (0.5 * norm((dxdt, dydt))**2)) * dydt - gradU[1]
+        return np.concatenate([dxdt, dydt, dvxdt, dvydt])
+
+    ## This solves an initial value problem to compute the next x,y coordinates
+    # This solves for the next x,y coordinates using the equations for dx/dt, dy/dt, dvx/dt, and dvy/dt and the previous
+    # x,y coordinates and velocitites as initial values. If the new x,y coordinates are outside the boundary, the
+    # reflect function for boundary is called
```

Comment:
Second sentence not needed. Should mention which solver is used, and maybe a reference to information about the solver.

Reply:
"the reflect function ... " -> "the points are reflected inside the domain using an elastic collision model. "